### ISSUE TITLE
Error comparing large sets of textual values from queries

### ISSUE THREAD

**Issue by daj102** [Seddryck/NBi#186]
Hi again,

So I'm having a really weird problem with NBi comparing two query result sets.  My system is a data warehouse with a staging database and a warehouse database, and we're writing tests to make sure that after we execute the transformation, we have the same keys in the warehouse as we found in the source data.

I'm trying to compare a very simple result set consisting of a textual key and a state.  The key _usually_ looks numeric, but there are some values that are really textual.  Examples of the result set are:

```
1  Suit
2  Claim
3  Suit
...
100001 Suit
...
E34F123 Claim
```

Now, if I define the query to sort on the first (textual) column, I get this result set:

```
1  Suit
10 Suit
100 Suit
1000 Claim
10000 Suit
...
101996 Suit
102  Claim
1020 Suit
```

You get the idea - I think the problem revolves somewhere around the fact that the length of the string can reduce from the previous value; not really sure because I can't give you a consistent test.

Anyway, if I sort by the first column, I can get up to around 2000 rows before it reports errors.  If I sort by the text length of the first column (I wondered if it was something to do with the way you were building the comparison tree) then i only get to the 19th row before it gives errors.  Even weirder, the test reports that the rows are both unexpected _and_ missing:
## Unexpected rows:

ResultSet with 1277 rows

| KEY (Text) | KEY (Text) | KEY (Numeric)  (+/- 0) |
| --- | --- | --- |
| 19 | Covered | 2 |
| 25 | Covered | 2 |
| 38 | Covered | 2 |
| 56 | Covered | 2 |
| 66 | Covered | 2 |
| 68 | Covered | 2 |
| 70 | Covered | 2 |
| 72 | Covered | 2 |
| 89 | Covered | 2 |
| 90 | Covered | 2 |

1267 (of 1277) rows have been skipped for display purpose.
## Missing rows:

ResultSet with 1277 rows

| KEY (Text) | KEY (Text) | KEY (Numeric)  (+/- 0) |
| --- | --- | --- |
| 19 | Covered | 2 |
| 25 | Covered | 2 |
| 38 | Covered | 2 |
| 56 | Covered | 2 |
| 66 | Covered | 2 |
| 68 | Covered | 2 |
| 70 | Covered | 2 |
| 72 | Covered | 2 |
| 89 | Covered | 2 |
| 90 | Covered | 2 |

1267 (of 1277) rows have been skipped for display purpose.

So ... I'm kind of at a loss as to what the possible cause may be.  It looks like, from some of the error messages I've seen while investigating this, that you are hashing the key text values ... but even if you were getting duplicate hashes, I wouldn't expect to see the rows reported as missing too.  I even tried casting the values to the exact same data type too, in case that had any effect on the hashing - but no.  I've also tried changing the definition of the columns (key/ignore/value) and data types (numeric/text) and I can still produce this error, albeit at different points in the dataset.

I've just downloaded the latest release (1.13.0.8) and tried it with that and it gives me the same behavior.  If this doesn't give you enough info, I can try and create a cut-down database for you to play with - just let me know.

Any ideas?

Test:

```xml
    <test name="CoverageMatchDescription">
        <system-under-test>
            <execution>
                <query connectionString="Data Source=myserver;Initial Catalog=Staging;Integrated Security=True;Pooling=False">
                    <![CDATA[
SELECT [ClaimLineBusinesskey] = CONVERT(VARCHAR(20), ln.G_Line)
    , [CovMatchDescription] = CONVERT(VARCHAR(20), CovMatDesc.T_VAL)
    , [len] = len(ln.G_Line)
FROM Staging.[Claims].[dbo_POLICY] p
INNER JOIN Staging.[Claims].[dbo_CLAIM] c
    ON c.G_POL = p.G_POL
        AND c.B_RCD_DEL = 0
INNER JOIN Staging.[Claims].[dbo_LINE] ln
    ON c.G_CLM = ln.G_CLM
        AND ln.B_RCD_DEL = 0
LEFT JOIN Staging.[Config].[dbo_CAT_KEY_LCL] CovMatDesc
    ON ln.C_COV_MATCH_STATUS = CovMatDesc.C_KEY
        AND CovMatDesc.N_APP = 'CLAIMS'
        AND CovMatDesc.N_CAT = '260'
ORDER BY 3,1        
                    ]]>
                </query>
            </execution>
        </system-under-test>
        <assert>
            <equalTo>
                <column index="0" role="key" type="text"/>
                <column index="1" role="key" type="text"/>
                <column index="2" role="key" type="numeric"/>
                <query connectionString="Data Source=myserver;Initial Catalog=EDW;Integrated Security=True;Pooling=False">
                     <![CDATA[
SELECT ClaimLineBusinessKey = CONVERT(VARCHAR(20), [ClaimLineBusinessKey])
    , [CovMatchDescription] = CONVERT(VARCHAR(20), coveragematchdescription)
    , [len] = len([ClaimLineBusinessKey])
FROM EDW.claims.Dim_Claimline
WHERE claimLinekey <> - 1
    AND RowIsCurrent = 'Y'
ORDER BY 3,1
                    ]]>
                </query>
            </equalTo>
        </assert>
    </test> 
```

**Comment by Seddryck** [comment]
Previously, there was a bug in NBi about hashing. We were hashing twice the keys (and bad implementation of IEquatable) ... this resulted in some result-sets complaining about duplicated rows when there was no duplicate in the result-set. But this bug has been fixed months ago and I've not touch this part since a long time. I never saw a false negative as the result of this failure.

When you say 

> If I sort by the first column, I can get up to around 2000 rows before it reports errors. If I sort by the text length of the first column (I wondered if it was something to do with the way you were building the comparison tree) then i only get to the 19th row before it gives errors. 

Do you mean that the count of unexpected and missing rows will vary depending on your order clause? That's indeed really weird.

What happens if you add a filter `WHERE CONVERT(VARCHAR(20), [ClaimLineBusinessKey])='19'` to both queries?

**Comment by daj102** [comment]
I get the same issue, but only for that row:

NBi.NUnit.Runtime.TestSuite.CoverageMatchDescription:
NBi.NUnit.Runtime.CustomStackTraceAssertionException : Execution of the query doesn't match the expected result 

  Expected: 
ResultSet with 1 row

| KEY (Text) | KEY (Text) |
| --- | --- |
| 19 | Covered |

  But was:  
ResultSet with 1 row

| KEY (Text) | KEY (Text) |
| --- | --- |
| 19 | Covered |

## Unexpected rows:

ResultSet with 1 row

| KEY (Text) | KEY (Text) |
| --- | --- |
| 19 | Covered |

## Missing rows:

ResultSet with 1 row

| KEY (Text) | KEY (Text) |
| --- | --- |
| 19 | Covered |

It's bizarre ... definitely have the right build:

10:56:01.139 Info  [ 5] TestAssemblyBuilder: Loaded assembly NBi.NUnit.Runtime, Version=1.13.0.8, Culture=neutral, PublicKeyToken=null

and I'm using .net 4.0:

10:56:00.623 Debug [ 1] RuntimeFrameworkSelector: Test will use net-4.0 framework

**Comment by Seddryck** [comment]
Converting a char to varchar will keep whitespace at the end. Your problem is probably there. Could you keep the filter on "19" and try to:
- change the settings to have numeric value for the first column and the second column will be a text but a value not a key
- switch back to your initial settings and remove all your convert(varchar(20), ...). 

I'll bet 1 cent that at least one of these trials will fix the issue.

**Comment by daj102** [comment]
So I started playing around with the data types and found that if I set _both_ sides to CHAR(20) instead of VARCHAR(20) then the test passes on the full data set.  I thought I might be looking like an idiot and your comment about the trailing spaces was something I missed ... but I added a function for LEN and tried replacing all spaces with a real character to force it to show up as different ... but it shows the correct values every time.  It's something about the conversion to VARCHAR, as you said.

I tried the following scenarios (L/R are the left/right sides of the comparison) with only one textual key column in the query, and one textual value column:

L = NCHAR(20), R = NCHAR(20) : PASS
L = NVARCHAR(20), R = NCHAR(20) : FAIL, all rows mismatch
L = NCHAR(20), R = NVARCHAR(20) : FAIL, all rows mismatch
L = NVARCHAR(20), R = NVARCHAR(20) : FAIL, 8745 out of 151993 rows fail

Interestingly, I then tried that last case but with a filter for ISNUMERIC() = 1 ... but it failed with this:

NBi.NUnit.Runtime.TestSuite.CoverageMatchDescription:
NBi.NUnit.Runtime.CustomStackTraceErrorException : In the column with index '0', NBi can't convert the value '0' to the type '4454072524512D93'. Key columns must match with their respective types and don't support null, generic or interval values.

So there's a bug in the IsNumeric function then ... :-)  I thought maybe that was the problem, but if I filter out any values with a letter, it still fails in text mode (although it passes in numeric mode).

TL;DR - you are right about the VARCHAR conversion, and **this isn't a bug in NBi**.  I found that someone changed the ANSI settings so that the spaces at the end of the CHAR were coming through to the VARCHAR (I usually have the setting to trim those off), and although the LEN function was showing the same number on both sides, one side had spaces on.

Thank you again!

**Comment by Seddryck** [comment]
Oki, thx for feedback
